spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/DistributionAndOrderingUtils.scala

```diff
@@ -80,7 +94,11 @@ object DistributionAndOrderingUtils {
       // the conversion to catalyst expressions above produces SortOrder expressions
       // for OrderedDistribution and generic expressions for ClusteredDistribution
       // this allows RepartitionByExpression to pick either range or hash partitioning
-      RepartitionByExpression(distribution, query, numShufflePartitions)
+      if (Spark3VersionUtil.isSpark30) {
+        repartitionByExpressionCtor.newInstance(distribution.toSeq, query, new Integer(numShufflePartitions))
+      } else {
+        repartitionByExpressionCtor.newInstance(distribution.toSeq, query, Some(numShufflePartitions))
```

Comment:
One thing that may out of the PR's scope. There is a feature introduced in SPARK-32056, iff `numShufflePartitions` is absent, partition coalesces for AQE will effect, which will avoid writing small files when AQE is enabled.

Reply:
Indeed, SPARK-32056, introduced in Spark 3.1, is what changed the signature so that the last parameter is an `Option[Int]` instead of an `Int`. If you look at the code change for SPARK-32056, if the `Option[Int]` is `None`, then the number of partitions is obtained from the conf:
```
  val numPartitions = optNumPartitions.getOrElse(SQLConf.get.numShufflePartitions)
```
And here, in `DistributionAndOrderingUtils.prepareQuery`, `numShufflePartitions` is indeed `conf.numShufflePartitions` from the `SQLConf` that is passed in.
So there is no reason to change the code here.

Reply:
Things are more interesting here, the `Strategy` also changed in SPARK-32056.
https://github.com/apache/spark/pull/28900/files#diff-21f071d73070b8257ad76e6e16ec5ed38a13d1278fe94bd42546c258a69f4410R687-R690

Reply:
I can see that there will be opportunities to vary the logic depending on Spark 3 version to take advantage of new Spark 3 features. However, as you remarked, that is out of the scope of this PR. Here we just want to support building and running against both 3.0 and 3.1 and keep the logic the same.

Reply:
I'm also +1 for keeping the logic the same in this PR, just wanna know if it will introduce side effects when set `numShufflePartitions` to `None`. @RussellSpitzer @aokolnychyi would you please take a look at it?

Reply:
I'm +1, I think as a follow up we should be checking if we are in Spark3 and doing something like
```
val numPartitions = write.requiredNumPartitions()
        val finalNumPartitions = if (numPartitions > 0) {
          numPartitions
        } else {
          conf.numShufflePartitions
        }
```

Which is what we are doing, but I think that's a bit of a perf improvement and not neccessary for getting this compatibility in. Maybe @wypoon you want to make a followup?

Reply:
@RussellSpitzer, what is the `write` in your code example?
I can look into a followup.

Reply:
Ah sorry we pass through the whole write (instead of just the distribution and ordering) as a parameter.

Reply:
@aokolnychyi is busy at the moment but I believe he'll want to put in some of the code changes to make this easier